native-image-config/build.gradle.kts

```diff
@@ -0,0 +1,335 @@
+        copyTypeConfig("serialization-config.json") { it["types"] }
+        copyProxyTypeConfig("proxy-config.json")
+        copyResourceConfig("resource-config.json")
+    }
+
+    private fun copyTypeConfig(fileName: String, typeArraySelector: (JsonNode) -> JsonNode = { it }) {
+        val config = loadJson(fileName)
+        filterTypes(fileName, typeArraySelector(config))
+        mapper.writeValue(targetDir.file(fileName).asFile, config)
+    }
+
+    private fun copyProxyTypeConfig(fileName: String) {
+        val config = loadJson(fileName)
+        filterProxyTypes(fileName, config)
+        mapper.writeValue(targetDir.file(fileName).asFile, config)
+    }
+
+    private fun copyResourceConfig(fileName: String) {
+        val config = loadJson(fileName)
+        val baseConfig = mapper.readTree(_baseResourceConfigFile.asFile)
+        val baseResourceRegexes = baseConfig.get("resources").get("includes").map {
+            val pattern = it.get("pattern").asText()
+            pattern.substringAfter(':', pattern).toRegex()
+        }.toList()
+        filterResources(fileName, config, baseResourceRegexes)
+        mapper.writeValue(targetDir.file(fileName).asFile, config)
+    }
+
+    private fun loadJson(fileName: String): JsonNode = mapper.readTree(sourceDir.file(fileName).asFile)
+
+    private fun filterTypes(fileName: String, array: JsonNode) {
+        requireArray(fileName, array)
+        val iterator = array.iterator()
+        while (iterator.hasNext()) {
+            val element = requireObject(fileName, iterator.next())
+            val typeName = getTypeName(element["name"].asText()) ?: continue
+            if (_excludedTypeRegexes.any { it.find(typeName) != null }) {
+                logger.info("$fileName: Excluding $typeName")
+                iterator.remove()
+                continue
+            }
+        }
+    }
+
+    private fun filterProxyTypes(fileName: String, array: JsonNode) {
+        requireArray(fileName, array)
+        val iterator = array.iterator()
+        while (iterator.hasNext()) {
+            val element = iterator.next()
+            val interfaceList = requireArray(fileName, element["interfaces"])
+            val interfaceIterator = interfaceList.iterator()
+            while (interfaceIterator.hasNext()) {
+                val interfaceElement = interfaceIterator.next()
+                val typeName = interfaceElement.asText()
+                if (_excludedTypeRegexes.any { it.find(typeName) != null }) {
+                    logger.info("$fileName: Excluding $typeName")
+                    interfaceIterator.remove()
+                }
+            }
+
+            if (interfaceList.isEmpty) {
+                iterator.remove()
+            }
+        }
+    }
+
+    private val resourceRegex = """^(?:[._a-zA-Z0-9:]*)\\Q(.*)\\E$""".toRegex()
+
+    private fun filterResources(fileName: String, obj: JsonNode, baseResourcePatterns: Iterable<Regex>) {
```

Comment:
Question) Assuming I would like to add a resource regex, how do I decide where to add it between `resource-config.json` and `excludedResourceRegexes`?

Reply:
You're correct, but they are meant for being used when our 'user' builds a native image. We just use the same set of rules to filter out irrelevant resource configurations in our native image config.